### PR TITLE
feat(jwt-audit-analytics-writer): add new fields to audit, DPoP details

### PR DESCRIPTION
**Pull request by alvimrtouch**
## Jira Issue

[PIN-10094](https://pagopa.atlassian.net/browse/PIN-10094)

## Context/Why
New fields have been added to the jwt audits. We want to persist them.

These are the new fields: type header, voucher's cnf claim, voucher's digest and client assertion's digest.

Moreover we have the DPoP information which is also supposed to be saved in a new table `jwt.dpop_audit`

## Services Impacted

- jwt-audit-analytics-writer
- models
- SQL tables

## Key Changes

- Added new DPoP table
- Persist DPoP informations to the new table
- Persist the new fields for jwt audits
- Fixed tests
- Add a skip for dpop saves when the dpop field is not present

## Traceability Checklist
- [X] Branch name contain the Jira key
- [X] PR title is compliant with the standard (type(scope): descrizione (KEY))

[PIN-10094]: https://pagopa.atlassian.net/browse/PIN-10094

### CODE REVIEW

File: packages/jwt-audit-analytics-writer/src/model/db.ts [hunk 1]
```diff
@@ -42,11 +42,41 @@ export const GeneratedTokenSchema = z.object({
   expiration_time: z.number().int(),
   expiration_time_tz: z.date(),
   issuer: z.string(),
+  cnf_jkt: z.string().optional(),
```
Comment (lorenzo-pa) [inline packages/jwt-audit-analytics-writer/src/model/db.ts:50]: It seems `typ` was added to both the SQL DDL (between `key_id` and `audience` in `jwt.generated_token_audit`) and to the domain model (`GeneratedTokenAuditDetails.typ: z.string()`), but it is missing from `GeneratedTokenSchema` here. It should appear between key_id and audience (earlier in the schema, before the diff range shown here).

INSERT mode: `typ` will always be NULL in the staging table — the field is silently dropped from every row.

COPY mode (critical): the CSV emitted by `generatedTokenMapping` has no `typ` column. Redshift COPY uses column position, so all fields from `audience` onward are shifted one slot left — `audience` value goes into `typ`, `subject` into `audience`, `not_before` into `subject`, etc., corrupting the entire row.

**Fix suggested: add typ: z.string().optional() to GeneratedTokenSchema between key_id and audience.**

File: packages/jwt-audit-analytics-writer/src/repositories/generatedToken.repository.ts
Comment (eduwardo) [inline packages/jwt-audit-analytics-writer/src/repositories/generatedToken.repository.ts:36]: `typ` is missing from the generated token mapping.

The upstream producer emits `typ` as `generatedToken.header.typ`, and the DB column was added after `key_id`. Since COPY ingestion is positional, skipping this field shifts all following CSV columns.

```suggestion
  key_id: (record) => record.keyId,
  typ: (record) => record.typ,
```

File: packages/jwt-audit-analytics-writer/src/repositories/generatedToken.repository.ts [hunk 2]
```diff
@@ -41,7 +41,11 @@ export const generatedTokenMapping: GeneratedTokenMapping = {
   expiration_time: (record) => record.expirationTime,
   expiration_time_tz: (record) => new Date(record.expirationTime),
   issuer: (record) => record.issuer,
+  cnf_jkt: (record) => record.cnf?.jkt,
```
Comment (lorenzo-pa) [inline packages/jwt-audit-analytics-writer/src/repositories/generatedToken.repository.ts:47]: Same issue as in `GeneratedTokenSchema`: `typ` is mapped in the domain model and present in the SQL DDL, but missing from `generatedTokenMapping`. This is the mapping that drives both the CSV column order (COPY mode) and the pg-promise column set (INSERT mode). Without it, COPY mode produces a column shift starting at audience.

File: packages/jwt-audit-analytics-writer/test/jwtAuditService.test.ts [hunk 3]
```diff
@@ -144,6 +152,9 @@ describe("JWT Audit Service tests", () => {
       );
       expect(generatedTokenStagingCount).toBe(0);
 
+      const dpopCount = await getTargetTableCount(conn, JwtDbTable.dpop);
+      expect(dpopCount).toBe(10);
```
Comment (lorenzo-pa) [inline packages/jwt-audit-analytics-writer/test/jwtAuditService.test.ts]: This test covers the happy path only: 10 records all with `dpop` → 10 DPoP rows in the target table. 
The PR checklist says "Add a skip for dpop saves when the dpop field is not present", but that guard (`filter((msg) => msg.dpop) in the service, return early in dpop.repository.ts:insert`) is never exercised by any test.

File: packages/jwt-audit-analytics-writer/test/utils.ts [hunk 4]
```diff
@@ -122,6 +128,20 @@ export const getMockJwtAudit = (): GeneratedTokenAuditDetails => {
       issuer: consumerId,
       expirationTime: new Date().getMilliseconds(),
     },
```
Comment (lorenzo-pa) [inline packages/jwt-audit-analytics-writer/test/utils.ts:138]: `getMockJwtAudit` was correctly updated with `typ`, `cnf`, and `dpop`, but the `digest` field (voucher's digest) is missing. `GeneratedTokenAuditDetails.digest` is now mapped to `digest_alg/digest_val` in `generatedTokenMapping`, so those columns exist in the staging table — but since no mock populates `digest`, every test is asserting with `digest_alg = null` and `digest_val = null`. The mapping code for those columns is never exercised with real values.

Per project convention: when new fields are added to the domain model, the mock factory must be updated to populate them. 
Suggestion: `digest: { alg: "SHA-256", value: "abc123..." }`,
alongside the other new fields, and add an assertion in at least one DB service test that `digest_alg` and `digest_val` are non-null after staging insertion.

File: docker/postgres/init-db.sql
Comment (eduwardo) [inline docker/postgres/init-db.sql:43]: The issue explicitly requires persisting the client assertion `digest` claim.

The upstream producer emits it as `clientAssertion.digest`, separately from the top-level generated token `digest`, so `client_assertion_audit` should include dedicated columns for it.



```suggestion
    expiration_time_tz TIMESTAMPTZ NOT NULL,
    digest_alg VARCHAR(50),
    digest_val VARCHAR(255),
```

File: packages/jwt-audit-analytics-writer/src/model/domain/models.ts [hunk 5]
```diff
@@ -18,11 +31,49 @@ export const ClientAssertionAuditDetails = z.object({
   subject: ClientId,
   audience: z.string(),
   expirationTime: z.number(),
+  digest: ClientAssertionDigest.optional(),
```
Comment (lorenzo-pa) [inline packages/jwt-audit-analytics-writer/src/model/domain/models.ts:34]: The PR description says "voucher's digest and client assertion's digest" should be persisted. The voucher's digest (`GeneratedTokenAuditDetails.digest`) is correctly mapped to `digest_alg/digest_val` in `generatedTokenMapping`. But `ClientAssertionAuditDetails.digest` added here has no corresponding columns in `client_assertion_audit` and no entry in `clientAssertionMapping` — it is parsed from the Kafka message and then silently discarded.

Is this intentional (only the voucher digest is stored, not the client assertion's)? If so, please remove this field from `ClientAssertionAuditDetails` to avoid confusion. If it should also be persisted, `digest_alg/digest_val` columns need to be added to `client_assertion_audit`, the SQL DDL, and `clientAssertionMapping`.

File: packages/jwt-audit-analytics-writer/src/model/db.ts
Comment (eduwardo) [inline packages/jwt-audit-analytics-writer/src/model/db.ts:39]: `GeneratedTokenSchema` should include `typ`, matching the `generated_token_audit` column order. This schema drives the merge column list, so the new DB column is currently not included in the target insert.

```suggestion
  key_id: z.string(),
  typ: z.string(),
```

Comment (eduwardo) [inline packages/jwt-audit-analytics-writer/src/model/db.ts:14]: `clientAssertion.digest` is part of the audit payload required by the issue, but it is missing from `ClientAssertionSchema`.

```suggestion
  expiration_time_tz: z.date(),
  digest_alg: z.string().optional(),
  digest_val: z.string().optional(),
```

File: docker/postgres/init-db.sql
Comment (eduwardo) [reply docker/postgres/init-db.sql:43]: These fields must be added to `clientAssertionMapping` as well.

```diff
diff --git a/packages/jwt-audit-analytics-writer/src/repositories/clientAssertion.repository.ts b/packages/jwt-audit-analytics-writer/src/repositories/clientAssertion.repository.ts
@@
   expiration_time_tz: (record) =>
     new Date(
       normalizeTimestampToMilliseconds(record.clientAssertion.expirationTime)
     ),
+  digest_alg: (record) => record.clientAssertion.digest?.alg,
+  digest_val: (record) => record.clientAssertion.digest?.value,
   generated_token_jwt_id: (record) => record.jwtId,
```

File: docker/postgres/init-db.sql [hunk 6]
```diff
@@ -45,6 +50,26 @@ CREATE TABLE IF NOT EXISTS jwt.client_assertion_audit (
...
+    typ VARCHAR(50) NOT NULL,
+    alg VARCHAR(50) NOT NULL,
+    jwk_kty VARCHAR(10) NOT NULL,
+    jwk_n TEXT,
+    jwk_e TEXT,
+    jwk_crv VARCHAR(10),
+    jwk_x TEXT,
+    jwk_y TEXT,
+    htm VARCHAR(16) NOT NULL,
+    htu TEXT NOT NULL,
```
Comment (eduwardo) [inline docker/postgres/init-db.sql]: `TEXT` is accepted by Redshift, but it is converted to `VARCHAR(256)`.

To keep the same pattern already used in this DDL, I would use explicit `VARCHAR(n)` sizes here as well.


```suggestion
    jwk_n VARCHAR(2048),
    jwk_e VARCHAR(64),
    jwk_crv VARCHAR(10),
    jwk_x VARCHAR(512),
    jwk_y VARCHAR(512),
    htm VARCHAR(16) NOT NULL,
    htu VARCHAR(1024) NOT NULL,
```

File: packages/jwt-audit-analytics-writer/src/model/domain/models.ts (hunk 5 above)
Comment (eduwardo) [reply packages/jwt-audit-analytics-writer/src/model/domain/models.ts:34]: Agree. I checked the upstream producer in `interop-be-monorepo`.

It emits two separate digest objects in the audit payload:

- `digest` at top level, coming from `generatedToken.payload.digest` -> voucher/generated token digest
- `clientAssertion.digest`, coming from `clientAssertion.payload.digest` -> client assertion digest

So yes, both should be persisted on KPI: the voucher digest on `generated_token_audit`, and the client assertion digest on `client_assertion_audit`. I added comments/suggestions below for the missing `client_assertion_audit` columns, `ClientAssertionSchema`, and `clientAssertionMapping`.

**Review by eduwardo — CHANGES_REQUESTED** [review]

File: packages/jwt-audit-analytics-writer/test/utils.ts (hunk 4 above)
Comment (alvimrtouch) [reply packages/jwt-audit-analytics-writer/test/utils.ts:138]: Added the new digest field to the mocks. Also added asserts for digest_alg and digest_val fields.

File: packages/jwt-audit-analytics-writer/test/jwtAuditService.test.ts (hunk 3 above)
Comment (alvimrtouch) [reply packages/jwt-audit-analytics-writer/test/jwtAuditService.test.ts]: Added a new assert to test the filtering.

**Review by eduwardo — APPROVED** [review]

**Review by lorenzo-pa — APPROVED** [review]
